File: csharp/uwapi/game.cs
namespace Unnatural
{
    using ConnectionStateEnum = Interop.UwConnectionStateEnum;
    using GameStateEnum = Interop.UwGameStateEnum;
    using MapStateEnum = Interop.UwMapStateEnum;

    public static class Game
    {
        public static void LogInfo(string msg)
        {
            Interop.uwLog(Interop.UwSeverityEnum.Info, msg);
        }

        public static void LogWarning(string msg)
        {
            Interop.uwLog(Interop.UwSeverityEnum.Warning, msg);
        }

        public static void LogError(string msg)
        {
            Interop.uwLog(Interop.UwSeverityEnum.Error, msg);
        }

        public static void SetPlayerName(string name)
        {
            Interop.uwSetPlayerName(name);
        }

        public static void PlayerJoinForce(uint forceId)
        {
            Interop.uwPlayerJoinForce(forceId);
        }

        public static void SetForceColor(float r, float g, float b) // [0 .. 1]
        {
            Interop.uwSetForceColor(r, g, b);
        }

        public static void SetForceRace(uint raceProto)
        {
            Interop.uwSetForceRace(raceProto);
        }

        public static void ForceJoinTeam(uint team)
        {
            Interop.uwForceJoinTeam(team);
        }

        public static void SkipCutscene()
        {
            Interop.uwSkipCutscene();
        }

        public static void SetConnectStartGui(bool startGui, string extraParams = "--observer 1")
        {
            Interop.uwSetConnectStartGui(startGui, extraParams);
        }

        public static bool ConnectFindLan(ulong timeoutMicroseconds = 1000000)
        {
            return Interop.uwConnectFindLan(timeoutMicroseconds);
        }

        public static void ConnectDirect(string address, ushort port)
        {
            Interop.uwConnectDirect(address, port);
        }

        public static void ConnectLobbyId(ulong lobbyId)
        {
            Interop.uwConnectLobbyId(lobbyId);
        }

        public static bool ConnectEnvironment()
        {
            return Interop.uwConnectEnvironment();
        }

        public static void ConnectNewServer(uint visibility = 0, string name = "", string extraParams = "")
        {
            Interop.uwConnectNewServer(visibility, name, extraParams);
        }

        // returns false if no data for reconnect are available
        // returns true after a reconnect was attempted (irrespective of whether actual connection has succeded)
        public static bool TryReconnect()
        {
            return Interop.uwTryReconnect();
        }

        public static void Disconnect()
        {
            Interop.uwDisconnect();
        }

        public static ConnectionStateEnum ConnectionState()
        {
            return Interop.uwConnectionState();
        }

        public static GameStateEnum GameState()
        {
            return Interop.uwGameState();
        }

        public static uint GameTick()
        {
            return Interop.uwGameTick();
        }

        public static void SetGameSpeed(float speed)
        {
            Interop.uwSetGameSpeed(speed);
        }

        public static void SetWeatherSpeed(float speed, float offset)
        {
            Interop.uwSetWeatherSpeed(speed, offset);
        }

        public static MapStateEnum MapState()
        {
            return Interop.uwMapState();
        }

        public static Interop.UwGameConfig GameConfig()
        {
            Interop.UwGameConfig data = new Interop.UwGameConfig();
            Interop.uwGameConfig(ref data);
            return data;
        }

        public static Interop.UwPerformanceStatistics PerformanceStatistics()
        {
            Interop.UwPerformanceStatistics data = new Interop.UwPerformanceStatistics();
            Interop.uwPerformanceStatistics(ref data);
            return data;
        }

        public static void PerformanceProfiling(bool enable)
        {
            Interop.uwPerformanceProfiling(enable);
        }

        public static ulong ProfilingEventBegin()
        {
            return Interop.uwProfilingEventBegin();
        }

        public static void ProfilingEventEnd(string name, ulong eventStartTime)
        {
            Interop.uwProfilingEventEnd(name, eventStartTime);
        }
    }
}

File: python/uwapi/game.py
from .interop import *


class Game:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def log_info(self, msg: str):
        uw_interop.uwLog(UwSeverityEnum.Info, msg)

    def log_warning(self, msg: str):
        uw_interop.uwLog(UwSeverityEnum.Warning, msg)

    def log_error(self, msg: str):
        uw_interop.uwLog(UwSeverityEnum.Error, msg)

    def set_player_name(self, name: str) -> None:
        uw_interop.uwSetPlayerName(name)

    def player_join_force(self, force_id: int) -> None:
        uw_interop.uwPlayerJoinForce(force_id)

    def set_force_color(self, r: float, g: float, b: float) -> None:
        uw_interop.uwSetForceColor(r, g, b)

    def set_force_race(self, race_proto: int) -> None:
        uw_interop.uwSetForceRace(race_proto)

    def force_join_team(self, team: int) -> None:
        uw_interop.uwForceJoinTeam(team)

    def skip_cutscene(self) -> None:
        uw_interop.uwSkipCutscene()

    def set_connect_start_gui(
        self, start_gui: bool, extra_params: str = "--observer 1"
    ) -> None:
        uw_interop.uwSetConnectStartGui(start_gui, extra_params)

    def connect_find_lan(self, timeout_microseconds: int = 1000000) -> bool:
        return uw_interop.uwConnectFindLan(timeout_microseconds)

    def connect_direct(self, address: str, port: int) -> None:
        uw_interop.uwConnectDirect(address, port)

    def connect_lobby_id(self, lobby_id: int) -> None:
        uw_interop.uwConnectLobbyId(lobby_id)

    def connect_environment(self) -> bool:
        return uw_interop.uwConnectEnvironment()

    def connect_new_server(
        self, visibility: int = 0, name: str = "", extra_params: str = ""
    ) -> None:
        uw_interop.uwConnectNewServer(visibility, name, extra_params)

    def try_reconnect(self) -> bool:
        return uw_interop.uwTryReconnect()

    def disconnect(self) -> None:
        uw_interop.uwDisconnect()

    def connection_state(self) -> UwConnectionStateEnum:
        return uw_interop.uwConnectionState()

    def game_state(self) -> UwGameStateEnum:
        return uw_interop.uwGameState()

    def game_tick(self) -> int:
        return uw_interop.uwGameTick()

    def map_state(self) -> UwMapStateEnum:
        return uw_interop.uwMapState()

    def performance_statistics(self) -> UwPerformanceStatistics:
        return uw_interop.uwPerformanceStatistics()

    def performance_profiling(self, enable: bool) -> None:
        uw_interop.uwPerformanceProfiling(enable)

    def profiling_event_begin(self) -> int:
        return uw_interop.uwProfilingEventBegin()

    def profiling_event_end(self, name: str, event_start_time: int) -> None:
        uw_interop.uwProfilingEventEnd(name, event_start_time)


uw_game = Game()
